File: strategies/templates/oq_template.js
/**
 * oq_template.js — Template strategy for /sphere quest (oq) — JavaScript.
 *
 * Copy this file to strategies/oq/<your_name>.js, rename the class, and
 * fill in the TODO sections.  Delete any methods you don't need — only
 * nextClick() is required.
 *
 * GAME OVERVIEW
 * -------------
 * Grid   : 5×5, all 25 cells start covered
 * Budget : 7 non-purple clicks
 * Goal   : find 3 of the 4 hidden purple spheres → the 4th converts to red
 *          (spR, 150 SP).  Collect red and spend remaining budget on tiles.
 *
 * Special click rules:
 *   Purple (spP) : click is FREE (does not consume a click)
 *   Red (spR)    : appears after 3 purples clicked; costs 1 click (worth 150 SP)
 *
 * Non-purple cells reveal the count of purple neighbours (orthogonal +
 * diagonal, capped at 4) as a color:
 *   spB = 0 purple neighbours
 *   spT = 1 purple neighbour
 *   spG = 2 purple neighbours
 *   spY = 3 purple neighbours
 *   spO = 4 purple neighbours
 *
 * COLOR REFERENCE (oq)
 * --------------------
 *   spP   purple  5 SP each; 4 hidden per board; click is FREE
 *   spR   red     150 SP; appears after 3 purples found; costs 1 click
 *   spB   blue    0 purple neighbours
 *   spT   teal    1 purple neighbour
 *   spG   green   2 purple neighbours
 *   spY   yellow  3 purple neighbours
 *   spO   orange  4 purple neighbours
 *
 * BOARD CELL FORMAT
 * -----------------
 * `board` is always an array of exactly 25 objects, one per cell:
 *   [{ row: number, col: number, color: string, clicked: boolean }, ...]
 * Row and col are 0-indexed (0..4).  color="spU" = covered/unknown.
 * clicked=false = still interactable; clicked=true = disabled.
 *
 * After 3 purples are clicked the 4th appears with color="spR", clicked=false.
 * Click it — it costs 1 click but is worth 150 SP.
 *
 * STATE MODEL
 * -----------
 * State lives inside the Node process — it is never sent back to the harness.
 *
 *   initEvaluationRun()          → run state (once before all games)
 *   initGamePayload(meta, rs)    → game state (reset before each game)
 *   nextClick(board, meta, gs)   → { row, col [, gameState] }
 *
 * initEvaluationRun(): return read-only data shared across all games.
 * initGamePayload():   return fresh per-game state (bridge resets it each game).
 * nextClick():         receive current game state, return { row, col } and
 *                      optionally a new gameState to replace it.
 *
 * meta keys (oq):
 *   clicks_left    number  remaining non-purple budget
 *   max_clicks     number  total non-purple budget (always 7)
 *   purples_found  number  purples clicked so far (0–3)
 *
 * See also
 * --------
 * oh/random_clicks.js   — minimal stateless example
 * oc/global_state.js    — initEvaluationRun() (global state) example
 * oq/stateful.js        — initGamePayload() + state threading (per-game state) example
 */

"use strict";

const { OQStrategy, register } = require("../../interface/strategy.js");

class MyOQStrategy extends OQStrategy {
  // -------------------------------------------------------------------------
  // Optional: global state (computed once, shared across ALL games)
  // -------------------------------------------------------------------------

  /**
   * Called ONCE before all games begin.
   *
   * Return anything that is board-independent and expensive to repeat.
   * The returned value is stored by the bridge and passed as `evaluationRunState`
   * to every initGamePayload() call.  Treat it as read-only.
   *
   * @returns {*}  Any value.  Default: null.
   */
  initEvaluationRun() {
    // TODO: replace with your global precomputation, or delete this method
    return null;
  }

  // -------------------------------------------------------------------------
  // Optional: per-game initialisation
  // -------------------------------------------------------------------------

  /**
   * Called once before each game's first click.
   *
   * @param {Object} meta   { clicks_left, max_clicks, purples_found }
   * @param {*}      evaluationRunState  Read-only value from initEvaluationRun()
   * @returns {*}   Initial gameState for this game's first nextClick() call.
   *
   * Example: initialise a constraint model — the set of all 12,650 possible
   * purple layouts, to be pruned as neighbor-count colors are revealed.
   */
  initGamePayload(meta, evaluationRunState) {
    // TODO: reset per-game fields here, or delete this method
    return evaluationRunState;
  }

  // -------------------------------------------------------------------------
  // Required: click decision
  // -------------------------------------------------------------------------

  /**
   * Choose the next cell to click.
   *
   * @param {Array<{row: number, col: number, color: string, clicked: boolean}>} board
   *   All 25 board cells.  After 3 purples are clicked, the 4th appears with
   *   color="spR", clicked=false.  Click it (costs 1 click, 150 SP).
   * @param {Object} meta
   *   { clicks_left: number, max_clicks: number, purples_found: number }
   * @param {*} gameState
   *   Value returned by the previous nextClick() (or initGamePayload() for first call).
   * @returns {{ row: number, col: number, gameState?: * }}
   *
   * Tips:
   *   - Click any cell with color="spR" and clicked=false — 150 SP for 1 click.
   *   - Click any cell with color="spP" and clicked=false immediately (free).
   *   - Each non-purple color gives a neighbour count for constraint inference.
   *   - Do not return a (row, col) where board[row*5+col].clicked is true.
   */
  nextClick(board, meta, gameState) {
    const clicked = new Set(board.filter(c => c.clicked).map(c => c.row * 5 + c.col));

    // Always click red when it appears — 150 SP for 1 click
    const reds = board.filter(c => c.color === "spR" && !c.clicked);
    if (reds.length > 0) return { row: reds[0].row, col: reds[0].col };

    // Always click purple immediately (free)
    const purples = board.filter(c => c.color === "spP" && !c.clicked);
    if (purples.length > 0) return { row: purples[0].row, col: purples[0].col };

    // TODO: replace the random fallback with your click logic
    const unclicked = [];
    for (let r = 0; r < 5; r++)
      for (let c = 0; c < 5; c++)
        if (!clicked.has(r * 5 + c)) unclicked.push([r, c]);

    if (unclicked.length === 0) return { row: 0, col: 0 };
    const [row, col] = unclicked[Math.floor(Math.random() * unclicked.length)];
    return { row, col };
  }
}

register(new MyOQStrategy());

File: strategies/templates/oq_template.py
"""Template strategy for /sphere quest (oq) — Python.

Copy this file to strategies/oq/<your_name>.py, rename the class, and fill
in the TODO sections.  Delete any methods you don't need — only next_click()
is required.

GAME OVERVIEW
-------------
Grid   : 5×5, all 25 cells start covered
Budget : 7 non-purple clicks
Goal   : find 3 of the 4 hidden purple spheres → the 4th converts to red
         (spR, 150 SP).  Collect red and spend remaining budget on high-value
         tiles.

Special click rules:
  - Purple (spP) : click is FREE (does not consume a click)
  - Red (spR)    : appears after 3 purples clicked; costs 1 click (worth 150 SP)

Non-purple cells reveal the count of purple neighbours (orthogonal +
diagonal, capped at 4) as a color:

  spB = 0 purple neighbours
  spT = 1 purple neighbour
  spG = 2 purple neighbours
  spY = 3 purple neighbours
  spO = 4 purple neighbours

This neighbor-count information can be used to narrow down where purples are
(similar to Minesweeper constraint inference).

COLOR REFERENCE (oq)
--------------------
spP   purple  5 SP each; 4 hidden per board; click is FREE
spR   red     150 SP; appears after 3 purples found; click is FREE
spB   blue    0 purple neighbours
spT   teal    1 purple neighbour
spG   green   2 purple neighbours
spY   yellow  3 purple neighbours
spO   orange  4 purple neighbours

BOARD CELL FORMAT
-----------------
board is always a list of exactly 25 dicts, one per cell:

    [{"row": int, "col": int, "color": str, "clicked": bool}, ...]

Row and col are 0-indexed (0..4).  color="spU" = covered/unknown.
clicked=False = still interactable; clicked=True = disabled.

After 3 purples are clicked the 4th purple appears with color="spR",
clicked=False in the board on the next call.  Click it — costs 1 click,
worth 150 SP.

META KEYS (oq)
--------------
clicks_left    int   remaining non-purple click budget
max_clicks     int   total non-purple click budget (always 7)
purples_found  int   number of purple cells clicked so far

STATE PAYLOAD
-------------
game_state is any Python object you choose.  The harness threads it through every
call within a game:

  init_evaluation_run()        → initial_state        (called once before all games)
  init_game_payload(meta, s0)    → s1                   (called once per game)
  next_click(..., s1)   → (row, col, s2)
  next_click(..., s2)   → (row, col, s3)
  ...

Use init_evaluation_run() for data computed ONCE and shared across all games
(lookup tables, precomputed weights, etc.).

Use init_game_payload() to reset per-game bookkeeping at the start of each game.

If your strategy is stateless, leave both optional methods out and return
`game_state` unchanged from next_click().

See also
--------
oh/random_clicks.py   — minimal stateless example
oc/global_state.py    — init_evaluation_run() (global state) example
oq/stateful.py        — init_game_payload() + state threading (per-game state) example
"""

import random
from typing import Any

from interface.strategy import OQStrategy


class MyOQStrategy(OQStrategy):
    # -----------------------------------------------------------------------
    # Optional: global state (computed once, shared across ALL games)
    # -----------------------------------------------------------------------

    def init_evaluation_run(self) -> Any:
        """Called once before the evaluation run begins.

        Compute anything that is board-independent and expensive to repeat.
        The returned value is passed as ``evaluation_run_state`` to init_game_payload() at the start
        of every game.  Treat it as read-only during play.

        Returns:
            Any Python object.  Default: None.

        Example: precompute for each of the 12,650 possible purple layouts
        the optimal first-click order, then look it up at runtime.
        """
        # TODO: replace with your global precomputation, or delete this method
        return None

    # -----------------------------------------------------------------------
    # Optional: per-game initialisation
    # -----------------------------------------------------------------------

    def init_game_payload(
        self,
        meta: dict[str, Any],
        evaluation_run_state: Any,
    ) -> Any:
        """Called once before each game's first click. Set up fresh per-game state.

        Args:
            meta:  game metadata (same keys as next_click).
            evaluation_run_state: read-only value from init_evaluation_run().

        Returns:
            The initial game_state for this game's first next_click() call.

        Example: initialise a constraint model — a set of all 12,650 possible
        purple layouts, narrowed down each turn as colors are revealed.
        """
        # TODO: reset per-game fields here, or delete this method
        return evaluation_run_state

    # -----------------------------------------------------------------------
    # Required: click decision
    # -----------------------------------------------------------------------

    def next_click(
        self,
        board: list[dict[str, Any]],
        meta: dict[str, Any],
        game_state: Any,
    ) -> tuple[int, int, Any]:
        """Choose the next cell to click.

        Args:
            board: all 25 board cells, each as
                   {"row": int, "col": int, "color": str, "clicked": bool}.
                   After 3 purples are clicked, the 4th appears with
                   color="spR", clicked=False — click it (costs 1 click, 150 SP).
            meta: {
                "clicks_left":   int,  # remaining non-purple budget
                "max_clicks":    int,  # total non-purple budget (always 7)
                "purples_found": int,  # purples clicked so far (0–3)
            }
            game_state: value returned by the previous next_click() call, or by
                   init_game_payload() for the first call of the game.

        Returns:
            (row, col, game_state)
            row, col    : 0-indexed coordinates of the cell to click next.
            game_state  : updated state to pass into the next call.

        Tips:
            - Click any cell with color="spR" and clicked=False — 150 SP for 1 click.
            - Click any cell with color="spP" and clicked=False immediately (free).
            - Each non-purple reveal gives a neighbor count — use for inference.
            - Do not return a (row, col) where board[row*5+col]["clicked"] is True.
        """
        clicked = {(c["row"], c["col"]) for c in board if c["clicked"]}

        # Always click red when it appears — 150 SP for 1 click
        reds = [(c["row"], c["col"]) for c in board if c["color"] == "spR" and not c["clicked"]]
        if reds:
            row, col = reds[0]
            return row, col, game_state

        # Always click purple immediately (free)
        purples = [(c["row"], c["col"]) for c in board if c["color"] == "spP" and not c["clicked"]]
        if purples:
            row, col = purples[0]
            return row, col, game_state

        # TODO: replace the random fallback with your click logic

        unclicked = [
            (r, c)
            for r in range(5)
            for c in range(5)
            if (r, c) not in clicked
        ]
        if not unclicked:
            return 0, 0, game_state
        row, col = random.choice(unclicked)
        return row, col, game_state
